clippy_lints/src/matches/manual_utils.rs

```diff
@@ -73,7 +75,7 @@ where
     }
 
     // `map` won't perform any adjustments.
-    if !cx.typeck_results().expr_adjustments(some_expr.expr).is_empty() {
+    if some_expr.expr_contains_ty_adjustments(cx) && expr_has_explicit_ty(cx, expr) {
         return None;
     }
```

Comment:
`expr_has_explicit_ty` misses some other [coercion sites](https://doc.rust-lang.org/stable/reference/type-coercions.html#coercion-sites) such as

```rust
fn main() {
    let x: &[u8; 4] = b"1234";
    f(match Some(()) {
        Some(_) => Some(Box::new(x)),
        None => None,
    });
}

fn f(_: Option<Box<&[u8]>>) {}
```

I think an easier way to rule out unrelated coercion false negatives would be to recursively check the coercion propagating subexpressions (see the second list in the above link) rather than every subexpression for adjustments

Reply:
sorry, shame to admit that I'm currently stuck on this 🤦‍♂️.

I'm pretty sure I misunderstand the concept, and I couldn't get it to work, especially for this specific test case:

```rust
//~v ERROR: manual implementation of `Option::map`
let _ = match Some(0) {
    Some(_) => Some(match f() {
        Ok(res) => Ok(Box::new(res)),
        _ => Err(()),
    }),
    None => None,
};
```

Do you mind provid some more guidence? 😅

Reply:
Looking into it more I was incorrect earlier, I had assumed that the point at which the adjustment happens wouldn't cross "past" a coercion site, i.e. in

```rust
let x: &[u8; 4] = b"1234";
f(match Some(()) {
    Some(_) => Some(Box::new(x)),
               ~~~~~~~~~~~~~~~~~ I thought the adjustment happened here
                             ~ but it happens here
    None => None,
});
```

I *think* checking for coercion propagating expressions **and** expressions that are coercion sites may still work though, roughly

```rust
fn is_coerced(expr) -> bool {
    if expr has adjustments {
        return true;
    }

    match expr.kind {
        // regular coercion sites
        Call | MethodCall => args.any(is_coerced),
        Struct | Enum | Union => fields.any(is_coerced),
        // propagating
        Array => elements.any(is_coerced),
        Repeat => is_coerced(repeated),
        Tuple => fields.any(is_coerced),
        Block => is_coerced(block.expr),
        If => is_coerced(then) || is_coerced(r#else),
        Match => arms.any(is_coerced),
    }
}
```

The key difference to the current implementation is the direction, we're looking at children rather than parents. In your example the adjustment that we don't care about happens at `f`, but we avoid a false negative because we never check that expression

```rust
let _ = match Some(0) {
    Some(_) => Some(match f() {
                          ~ adjustment here
        Ok(res) => Ok(Box::new(res)),
        _ => Err(()),
    }),
    None => None,
};
```

There will be false positives though:

```rust
fn g(b: &[u8]) -> Box<&[u8]> {
    Box::new(b)
}

let x: &[u8; 4] = b"1234";

// ok
let _: Option<Box<&[u8]>> = match Some(0) {
    Some(_) => Some(g(x)),
    None => None,
};

// also ok
let _: Option<Box<&[u8]>> = Some(0).map(|_| g(x));
```

You might be able to eliminate these by checking that the type of the relevant arg/field/etc is concrete, but also accepting those FNs is perfectly fine to me

Reply:
There is an interesting FN case in `manual_map_option.rs`:

```rust
match Some(&String::new()) {
    Some(x) => Some(str::len(x)),
    None => None,
};
```
the `&String` was coerced to `&str` but it shouldn't matter right? I don't know if checking the type for concrete concreteness is possible, thus I'm still finding a way to eliminate this 😸

Reply:
From a `Ty` I think it would be something like `.has_type_flags(TypeFlags::HAS_TY_PARAM)`

You can use [`expr_sig`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/ty/fn.expr_sig.html) on the callee/receiver as like a `tcx.fn_sig` that works for closures/etc too

If the output of the function doesn't have a ty param then it should(?) be fine to `return false`

For individual args you could zip the inputs with the args, only checking the arg if it `HAS_TY_PARAM`

Reply:
That works perfectly! Thank you so much 👍